In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

* Import du premier fichier item properties - REMPLACER PAR BON CHEMIN 


In [2]:
df_properties1 = pd.read_csv("/Volumes/HD 2/DataScientest/PROJET/Jeu de données/donnees_brutes/item_properties_part1.csv")
df_properties1.head()

timestamp  itemid    property                            value
0  1435460400000  460429  categoryid                             1338
1  1441508400000  206783         888          1116713 960601 n277.200
2  1439089200000  395014         400  n552.000 639502 n720.000 424566
3  1431226800000   59481         790                       n15360.000
4  1431831600000  156781         917                           828513

* Import du second fichier item properties - REMPLACER PAR BON CHEMIN 

In [3]:
df_properties2 = pd.read_csv("/Volumes/HD 2/DataScientest/PROJET/Jeu de données/donnees_brutes/item_properties_part2.csv")
df_properties2.head()

timestamp  itemid property            value
0  1433041200000  183478      561           769062
1  1439694000000  132256      976  n26.400 1135780
2  1435460400000  420307      921  1149317 1257525
3  1431831600000  403324      917          1204143
4  1435460400000  230701      521           769062

* Concaténation des 2 fichiers item properties dans df_properties car fichiers séparés cause nombre de lignes trop important

In [4]:
df_properties = pd.concat([df_properties1, df_properties2])
df_properties.head()

timestamp  itemid    property                            value
0  1435460400000  460429  categoryid                             1338
1  1441508400000  206783         888          1116713 960601 n277.200
2  1439089200000  395014         400  n552.000 639502 n720.000 424566
3  1431226800000   59481         790                       n15360.000
4  1431831600000  156781         917                           828513

* Affichage des infos de df_properties et recherche de valeurs nulles

In [5]:
df_properties.info()
df_properties.isna().sum()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20275902 entries, 0 to 9275902
Data columns (total 4 columns):
 #   Column     Dtype 
---  ------     ----- 
 0   timestamp  int64 
 1   itemid     int64 
 2   property   object
 3   value      object
dtypes: int64(2), object(2)
memory usage: 773.5+ MB


timestamp    0
itemid       0
property     0
value        0
dtype: int64

_Nous donne 20275902 lignes et aucune valeur nulle_

* Nombre valeurs uniques dans df_properties

In [6]:
df_properties.nunique(axis=0)

timestamp         18
itemid        417053
property        1104
value        1966868
dtype: int64

* Équilibre des valeurs sur df_properties timestamp (calcul pour compléter le fichier rapport d'exploration)

In [7]:
print(df_properties['timestamp'].count())
print(df_properties['timestamp'].unique())
print(df_properties['timestamp'].value_counts())

20275902
[1435460400000 1441508400000 1439089200000 1431226800000 1431831600000
 1436065200000 1434250800000 1433646000000 1439694000000 1437274800000
 1436670000000 1432436400000 1438484400000 1442113200000 1433041200000
 1440903600000 1440298800000 1437879600000]
1431226800000    3012756
1433041200000    2620101
1435460400000    2368972
1432436400000    1815598
1433646000000    1686681
1439694000000    1488244
1431831600000    1027889
1434250800000     755965
1436670000000     593657
1442113200000     583018
1440298800000     571339
1441508400000     566183
1437274800000     563575
1439089200000     542778
1440903600000     532586
1438484400000     529315
1437879600000     519460
1436065200000     497785
Name: timestamp, dtype: int64


* Équilibre des valeurs sur df_properties itemid (calcul pour compléter le fichier rapport d'exploration)

In [8]:
print(df_properties['itemid'].count())
print(df_properties['itemid'].unique())
print(df_properties['itemid'].value_counts())

20275902
[460429 206783 395014 ... 449778 352709 170360]
158903    468
254069    462
91855     461
150800    459
339536    444
         ... 
97687      15
342400     14
73456      13
207227     12
243157     12
Name: itemid, Length: 417053, dtype: int64


* Équilibre des valeurs sur df_properties property (calcul pour compléter le fichier rapport d'exploration)

In [9]:
print(df_properties['property'].count())
print(df_properties['property'].unique())
print(df_properties['property'].value_counts())

20275902
['categoryid' '888' '400' ... '1091' '522' '769']
888           3000398
790           1790516
available     1503639
categoryid     788214
6              631471
               ...   
472                 1
744                 1
769                 1
1046                1
530                 1
Name: property, Length: 1104, dtype: int64


* Équilibre des valeurs sur df_properties value (calcul pour compléter le fichier rapport d'exploration)

In [10]:
print(df_properties['value'].count())
print(df_properties['value'].unique())
print(df_properties['value'].value_counts())

20275902
['1338' '1116713 960601 n277.200' 'n552.000 639502 n720.000 424566' ...
 '1165718' '598729 511183 981281 829153 631983' 'n1020780.000']
769062                                                                                                                                                                                                                                     1537247
0                                                                                                                                                                                                                                           863246
1                                                                                                                                                                                                                                           641420
679677                                                                                                                        

* Import du  fichier des catégories - REMPLACER PAR BON CHEMIN

In [11]:
cat_tree = pd.read_csv("/Volumes/HD 2/DataScientest/PROJET/Jeu de données/donnees_brutes/category_tree.csv")
cat_tree.head()

categoryid  parentid
0        1016     213.0
1         809     169.0
2         570       9.0
3        1691     885.0
4         536    1691.0

* Infos de cat_tree et recherche de valeurs nulles

In [12]:
cat_tree.info()
cat_tree.isna().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1669 entries, 0 to 1668
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   categoryid  1669 non-null   int64  
 1   parentid    1644 non-null   float64
dtypes: float64(1), int64(1)
memory usage: 26.2 KB


categoryid     0
parentid      25
dtype: int64

_Nous donne 1669 rows ainsi que 25 valeurs nulles dans la colonne parentid_

* Pourcentage de valeurs nulles trouvées dans cat_tree :

In [13]:
na_parentid = (cat_tree.parentid.isna().sum() / cat_tree.shape[0]) * 100
print("parentid a", na_parentid, "% de na")

parentid a 1.4979029358897544 % de na


* Nombre valeurs uniques dans cat_tree :

In [14]:
cat_tree.nunique(axis=0)

categoryid    1669
parentid       362
dtype: int64

* Équilibre des valeurs sur cat_tree categoryid (calcul pour compléter le fichier rapport d'exploration)

In [15]:
print(cat_tree['categoryid'].count())
print(cat_tree['categoryid'].unique())
print(cat_tree['categoryid'].value_counts())

1669
[1016  809  570 ... 1336  689  761]
1697    1
1274    1
1120    1
1122    1
1124    1
       ..
577     1
579     1
581     1
583     1
0       1
Name: categoryid, Length: 1669, dtype: int64


* Équilibre des valeurs sur cat_tree parentid (calcul pour compléter le fichier rapport d'exploration)

In [16]:
print(cat_tree['parentid'].count())
print(cat_tree['parentid'].unique())
print(cat_tree['parentid'].value_counts())

1644
[ 213.  169.    9.  885. 1691.   nan  378.  542. 1537. 1621.  893.  312.
  622. 1281. 1110. 1604.  945.  768.  229. 1181.  711.   73.  393.  250.
 1420. 1101. 1424. 1215.  313.  583.   71.  590. 1550.  763.  488.  232.
 1591.  383.  949.    8.  880.  322.  666. 1291.  865.  186.  336.   92.
  585.  399.  178. 1682.  777.  520.  781. 1453.  571.  252.  866. 1669.
  630. 1370.  609. 1618.  351. 1486.  854. 1229. 1009. 1259. 1490. 1444.
  593.  732. 1394.  418. 1687. 1658.  216. 1251.  727.  810.  894.  480.
 1299. 1416.  561.  395. 1615.  654.  506.  900. 1278.  140.  653. 1330.
 1562. 1637. 1095.  384.  805. 1458. 1308. 1519. 1329.  149. 1239. 1667.
 1383.  540. 1515.  825.  121. 1534. 1371.  409.  131.  369. 1499.  886.
  920.  827.  346. 1136.   61.  897. 1058. 1430.  164.  314. 1202. 1532.
  293.  938.  986.  206.  766.  107. 1492. 1174.  937.  594.  207. 1497.
  658.  188. 1323. 1200. 1396. 1031. 1156.  947. 1012. 1306.  500.  226.
 1066.  538.  679. 1139.  724. 1008.  901. 159

* Import du fichier des événements, jeu de données comportementales - REMPLACER PAR BON CHEMIN



In [17]:
df_events = pd.read_csv("/Volumes/HD 2/DataScientest/PROJET/Jeu de données/donnees_brutes/events.csv")
df_events.head()

timestamp  visitorid event  itemid  transactionid
0  1433221332117     257597  view  355908            NaN
1  1433224214164     992329  view  248676            NaN
2  1433221999827     111016  view  318965            NaN
3  1433221955914     483717  view  253185            NaN
4  1433221337106     951259  view  367447            NaN

* Infos de df_events et recherche de valeurs nulles

In [18]:
df_events.info()
df_events.isna().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2756101 entries, 0 to 2756100
Data columns (total 5 columns):
 #   Column         Dtype  
---  ------         -----  
 0   timestamp      int64  
 1   visitorid      int64  
 2   event          object 
 3   itemid         int64  
 4   transactionid  float64
dtypes: float64(1), int64(3), object(1)
memory usage: 105.1+ MB


timestamp              0
visitorid              0
event                  0
itemid                 0
transactionid    2733644
dtype: int64

_Donne 2756101 rows et 2733644 valeurs nulels dans transaction id reflétant un nombre de transactions très bas_

* Nombre valeurs uniques dans df_events :

In [19]:
df_events.nunique(axis=0)

timestamp        2750455
visitorid        1407580
event                  3
itemid            235061
transactionid      17672
dtype: int64

_1407580 Utilisateurs unique,_
_235061 Produits,_
_17672 transactions_

* Nombre de lignes dupliquées (que dans df_events, aucune dans les autres dataframes)

In [20]:
df_events[df_events.duplicated()].shape

(460, 5)

_On constate 460 lignes dupliquées, on les supprime du dataframe df_events :_

In [21]:
df_events.drop_duplicates(inplace=True)

* Répartition du nombre de valeurs dans la colonne event

In [22]:
df_events.event.value_counts()

view           2664218
addtocart        68966
transaction      22457
Name: event, dtype: int64

_2664218 produits consultés,_
_68966 produits ajoutés au panier,_
_22457 produits vendus_

* Avec si peu de transactions on souhaite visualiser le pourcentage de valeurs nulles que cela représente :

In [23]:
df_events.isna().mean()

timestamp        0.000000
visitorid        0.000000
event            0.000000
itemid           0.000000
transactionid    0.991851
dtype: float64

_On a 99% de valeurs nulles sur transactionid_

* On souhaite regarder le nombre de valeurs non nulles quand l'événement n'est pas une transaction.

In [24]:
pd.crosstab(df_events['event'], df_events['transactionid'].isna())

transactionid  False    True 
event                        
addtocart          0    68966
transaction    22457        0
view               0  2664218

_Le 0 indique que la variable transactionid est manquante uniquement lorsque le type d'événement n'est pas une transaction_

* df_events est le dataframe qui va le plus nous intéresser pour des calculs d'événements
* Conversion du timestamp en dates réelles dans df_events (changement de l'unix en date et creation d'une colonne jour de la semaine (les heures n'avaient pas de sens donc décalage de 5h pour match le -5 UTC americain) :

In [25]:
import datetime
import calendar

df_events['timestamp'] = pd.to_datetime(df_events['timestamp'], unit= "ms",) - datetime.timedelta(hours=5)
df_events['day_of_week'] = df_events.timestamp.dt.day_name()

df_events.head()

timestamp  visitorid event  itemid  transactionid day_of_week
0 2015-06-02 00:02:12.117     257597  view  355908            NaN     Tuesday
1 2015-06-02 00:50:14.164     992329  view  248676            NaN     Tuesday
2 2015-06-02 00:13:19.827     111016  view  318965            NaN     Tuesday
3 2015-06-02 00:12:35.914     483717  view  253185            NaN     Tuesday
4 2015-06-02 00:02:17.106     951259  view  367447            NaN     Tuesday

* On remet les jours de la semaine dans l'ordre et on sépare les valeurs de la colonne event en 3 colonnes

In [26]:
cats = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
df_events['day_of_week'] = pd.Categorical(df_events['day_of_week'], categories=cats, ordered=True)

df_events = df_events.join(pd.get_dummies(df_events.event, prefix = "events"))
df_events.head()

timestamp  visitorid event  itemid  transactionid day_of_week  \
0 2015-06-02 00:02:12.117     257597  view  355908            NaN     Tuesday   
1 2015-06-02 00:50:14.164     992329  view  248676            NaN     Tuesday   
2 2015-06-02 00:13:19.827     111016  view  318965            NaN     Tuesday   
3 2015-06-02 00:12:35.914     483717  view  253185            NaN     Tuesday   
4 2015-06-02 00:02:17.106     951259  view  367447            NaN     Tuesday   

   events_addtocart  events_transaction  events_view  
0                 0                   0            1  
1                 0                   0            1  
2                 0                   0            1  
3                 0                   0            1  
4                 0                   0            1

* Durée data set df_events

In [27]:
df_events.timestamp.max() - df_events.timestamp.min()

Timedelta('137 days 23:59:43.404000')

_Soit une période de 4,5 mois_

* Max et min de la variable timestamp convertie dans df_events

In [28]:
df_events['timestamp'] = pd.to_datetime(df_events['timestamp'])
df_events['timestamp'].head()
maxi = df_events.max(axis=0)['timestamp']
mini = df_events.min(axis=0)['timestamp']
print('la date de debut du dataset est le', mini ,"\n",'et la date de fin du dataset est le', maxi)

la date de debut du dataset est le 2015-05-02 22:00:04.384000 
 et la date de fin du dataset est le 2015-09-17 21:59:47.788000


* Pourcentage de valeurs nulles dans df_events, colonne transaction :

In [29]:
na_transaction = (df_events.transactionid.isna().sum() / df_events.shape[0]) * 100
print("transactionid a ", na_transaction, "% de na")

transactionid a  99.18505349572023 % de na


* Équilibre des valeurs sur df_events timestamp (calcul pour compléter le fichier rapport d'exploration)

In [30]:
print(df_events['timestamp'].count())
print(df_events['timestamp'].unique())
print(df_events['timestamp'].value_counts())

2755641
['2015-06-02T00:02:12.117000000' '2015-06-02T00:50:14.164000000'
 '2015-06-02T00:13:19.827000000' ... '2015-07-31T21:57:00.527000000'
 '2015-07-31T22:08:50.703000000' '2015-07-31T22:36:03.914000000']
2015-05-15 01:39:17.755    6
2015-06-17 14:13:47.120    6
2015-06-02 20:24:40.140    6
2015-05-21 12:40:49.621    6
2015-06-17 14:07:07.808    6
                          ..
2015-09-07 12:58:07.288    1
2015-05-18 08:22:52.504    1
2015-08-11 16:13:12.646    1
2015-08-06 10:50:22.597    1
2015-06-03 09:59:36.772    1
Name: timestamp, Length: 2750455, dtype: int64


* Équilibre des valeurs sur df_events visitorid  (calcul pour compléter le fichier rapport d'exploration)

In [31]:
print(df_events['visitorid'].count())
print(df_events['visitorid'].unique())
print(df_events['visitorid'].value_counts())

2755641
[ 257597  992329  111016 ...  591435  762376 1184451]
1150086    7757
530559     4328
152963     3024
895999     2474
163561     2410
           ... 
918817        1
920864        1
1047838       1
1041693       1
619217        1
Name: visitorid, Length: 1407580, dtype: int64


* On constate un déséquilibre, le visitorid 1150086 est venu 7757 fois, le visitorid 530559 est venu 4328 fois, suspicion de bots
* Pour identifier s'il y a des bots on définie des bornes arbitrairement et on classe selon ce qu'on estime être des fréquences d'utilisateurs faibles à excessives. Au-dessus de 3 000, on classe ces visitorid dans botts

In [32]:
df_events['User frequency'] = df_events['visitorid'].map(df_events['visitorid'].value_counts())
borne = [0, 10, 50, 3000, 7757]
label = ["Infrequent user","Recurrent user","Big user","Bots"]
df_events["User category"] =  pd.cut(df_events['User frequency'],borne, labels = label)
df_events.head()

timestamp  visitorid event  itemid  transactionid day_of_week  \
0 2015-06-02 00:02:12.117     257597  view  355908            NaN     Tuesday   
1 2015-06-02 00:50:14.164     992329  view  248676            NaN     Tuesday   
2 2015-06-02 00:13:19.827     111016  view  318965            NaN     Tuesday   
3 2015-06-02 00:12:35.914     483717  view  253185            NaN     Tuesday   
4 2015-06-02 00:02:17.106     951259  view  367447            NaN     Tuesday   

   events_addtocart  events_transaction  events_view  User frequency  \
0                 0                   0            1               2   
1                 0                   0            1              30   
2                 0                   0            1               2   
3                 0                   0            1               3   
4                 0                   0            1               1   

     User category  
0  Infrequent user  
1   Recurrent user  
2  Infrequent user  
3  Infrequent user  
4  Infrequent user

* On compte les nombre de valeurs de ces nouvelles catégories. Cela nous révèle un nombre anormal de fréquences élevées

In [33]:
df_events['User category'].value_counts()

Infrequent user    2209275
Recurrent user      332705
Big user            198552
Bots                 15109
Name: User category, dtype: int64

* On fait la même manipulation sur les items pour mesurer la popularité de ces derniers 

In [34]:
df_events['Product frequency'] = df_events['itemid'].map(df_events['itemid'].value_counts())
borne = [0,100, 1000, 1500, 2000]
label = ["Unpopular products","Popular products","Very popular products","Most popular products"]
df_events['Product category'] =  pd.cut(df_events['Product frequency'],borne, labels = label)
df_events.head()

timestamp  visitorid event  itemid  transactionid day_of_week  \
0 2015-06-02 00:02:12.117     257597  view  355908            NaN     Tuesday   
1 2015-06-02 00:50:14.164     992329  view  248676            NaN     Tuesday   
2 2015-06-02 00:13:19.827     111016  view  318965            NaN     Tuesday   
3 2015-06-02 00:12:35.914     483717  view  253185            NaN     Tuesday   
4 2015-06-02 00:02:17.106     951259  view  367447            NaN     Tuesday   

   events_addtocart  events_transaction  events_view  User frequency  \
0                 0                   0            1               2   
1                 0                   0            1              30   
2                 0                   0            1               2   
3                 0                   0            1               3   
4                 0                   0            1               1   

     User category  Product frequency    Product category  
0  Infrequent user                 57  Unpopular products  
1   Recurrent user                 34  Unpopular products  
2  Infrequent user                 15  Unpopular products  
3  Infrequent user                 14  Unpopular products  
4  Infrequent user                260    Popular products

* Équilibre des valeurs sur df_events event  (calcul pour compléter le fichier rapport d'exploration)

In [35]:
print(df_events['event'].count())
print(df_events['event'].unique())
print(df_events['event'].value_counts())

2755641
['view' 'addtocart' 'transaction']
view           2664218
addtocart        68966
transaction      22457
Name: event, dtype: int64


* Équilibre des valeurs sur df_events itemid  (calcul pour compléter le fichier rapport d'exploration)

In [36]:
print(df_events['itemid'].count())
print(df_events['itemid'].unique())
print(df_events['itemid'].value_counts())

2755641
[355908 248676 318965 ...    613  52086 177353]
187946    3412
461686    2975
5411      2334
370653    1854
219512    1800
          ... 
214439       1
37800        1
227720       1
223626       1
2047         1
Name: itemid, Length: 235061, dtype: int64


In [37]:
df_events.head()

timestamp  visitorid event  itemid  transactionid day_of_week  \
0 2015-06-02 00:02:12.117     257597  view  355908            NaN     Tuesday   
1 2015-06-02 00:50:14.164     992329  view  248676            NaN     Tuesday   
2 2015-06-02 00:13:19.827     111016  view  318965            NaN     Tuesday   
3 2015-06-02 00:12:35.914     483717  view  253185            NaN     Tuesday   
4 2015-06-02 00:02:17.106     951259  view  367447            NaN     Tuesday   

   events_addtocart  events_transaction  events_view  User frequency  \
0                 0                   0            1               2   
1                 0                   0            1              30   
2                 0                   0            1               2   
3                 0                   0            1               3   
4                 0                   0            1               1   

     User category  Product frequency    Product category  
0  Infrequent user                 57  Unpopular products  
1   Recurrent user                 34  Unpopular products  
2  Infrequent user                 15  Unpopular products  
3  Infrequent user                 14  Unpopular products  
4  Infrequent user                260    Popular products

* Équilibre des valeurs sur df_events transactionid  (calcul pour compléter le fichier rapport d'exploration)

In [38]:
print(df_events['transactionid'].count())
print(df_events['transactionid'].value_counts())

22457
7063.0     31
765.0      28
8351.0     27
2753.0     23
6993.0     21
           ..
11740.0     1
11739.0     1
11738.0     1
11737.0     1
2.0         1
Name: transactionid, Length: 17672, dtype: int64


• Comptage des valeurs uniques dans le dataframe

In [39]:
a = len(df_events['itemid'].unique())
b = len(df_events['visitorid'].unique())
c = len(df_events['transactionid'].unique())
print('on observe', a,'produits uniques,', b,'visiteurs uniques et', c, 'transactions uniques dans le dataframe')

on observe 235061 produits uniques, 1407580 visiteurs uniques et 17673 transactions uniques dans le dataframe


* Observation de quelques "super-paniers" avec des paniers contenant de nombreux articles (ici test 28 articles, même ID de transaction)

In [40]:
df_events.loc[df_events['transactionid'] == 765.0]

timestamp  visitorid        event  itemid  \
1683146 2015-05-14 11:40:13.675     761482  transaction  193908   
1683181 2015-05-14 11:40:13.707     761482  transaction  178290   
1683280 2015-05-14 11:40:13.400     761482  transaction  406165   
1683295 2015-05-14 11:40:13.707     761482  transaction  169720   
1683298 2015-05-14 11:40:13.307     761482  transaction  447818   
1683367 2015-05-14 11:40:13.817     761482  transaction  131861   
1687066 2015-05-14 11:40:13.628     761482  transaction  389158   
1687071 2015-05-14 11:40:13.628     761482  transaction  265846   
1687186 2015-05-14 11:40:13.322     761482  transaction  256146   
1687206 2015-05-14 11:40:13.338     761482  transaction  130599   
1687253 2015-05-14 11:40:13.470     761482  transaction   39439   
1687258 2015-05-14 11:40:13.501     761482  transaction  185241   
1687260 2015-05-14 11:40:13.501     761482  transaction  310525   
1687302 2015-05-14 11:40:13.400     761482  transaction  386092   
1687317 2015-05-14 11:40:13.581     761482  transaction  224818   
1688108 2015-05-14 11:40:13.447     761482  transaction  302909   
1694809 2015-05-14 11:40:13.485     761482  transaction  296749   
1694990 2015-05-14 11:40:13.832     761482  transaction  438605   
1697078 2015-05-14 11:40:13.597     761482  transaction  367318   
1697144 2015-05-14 11:40:13.691     761482  transaction  137169   
1697302 2015-05-14 11:40:13.307     761482  transaction  108096   
1697332 2015-05-14 11:40:13.785     761482  transaction   82340   
1702432 2015-05-14 11:40:13.597     761482  transaction  191050   
1702451 2015-05-14 11:40:13.291     761482  transaction  262699   
1702455 2015-05-14 11:40:13.338     761482  transaction  265775   
1702460 2015-05-14 11:40:13.470     761482  transaction  276886   
1702467 2015-05-14 11:40:13.566     761482  transaction  380997   
1702468 2015-05-14 11:40:13.848     761482  transaction   41803   

         transactionid day_of_week  events_addtocart  events_transaction  \
1683146          765.0    Thursday                 0                   1   
1683181          765.0    Thursday                 0                   1   
1683280          765.0    Thursday                 0                   1   
1683295          765.0    Thursday                 0                   1   
1683298          765.0    Thursday                 0                   1   
1683367          765.0    Thursday                 0                   1   
1687066          765.0    Thursday                 0                   1   
1687071          765.0    Thursday                 0                   1   
1687186          765.0    Thursday                 0                   1   
1687206          765.0    Thursday                 0                   1   
1687253          765.0    Thursday                 0                   1   
1687258          765.0    Thursday                 0                   1   
1687260          765.0    Thursday                 0                   1   
1687302          765.0    Thursday                 0                   1   
1687317          765.0    Thursday                 0                   1   
1688108          765.0    Thursday                 0                   1   
1694809          765.0    Thursday                 0                   1   
1694990          765.0    Thursday                 0                   1   
1697078          765.0    Thursday                 0                   1   
1697144          765.0    Thursday                 0                   1   
1697302          765.0    Thursday                 0                   1   
1697332          765.0    Thursday                 0                   1   
1702432          765.0    Thursday                 0                   1   
1702451          765.0    Thursday                 0                   1   
1702455          765.0    Thursday                 0                   1   
1702460          765.0    Thursday                 0                   1   
1702467          765.0    

* Produits les plus populaires (vue, panier, transaction confondues) sur df_events

In [41]:
top_produits = df_events['itemid'].value_counts()
top_produits.sort_values(ascending=False)
top_produits.head(10)

187946    3412
461686    2975
5411      2334
370653    1854
219512    1800
257040    1647
298009    1642
96924     1633
309778    1628
384302    1608
Name: itemid, dtype: int64

* Produits les moins populaires (vue, panier, transaction confondues) sur df_events

In [42]:
top_produits.tail(10)

317820    1
218533    1
211328    1
207234    1
319096    1
214439    1
37800     1
227720    1
223626    1
2047      1
Name: itemid, dtype: int64

* On souhaite réaliser l'étude statistique d'une variable cible (events_transaction) avec d'autres variables quantitatives afin d'analyser la corrélation inter-variables :

In [43]:
from scipy.stats import pearsonr
pd.DataFrame(pearsonr(df_events['events_transaction'],df_events['User frequency']), index = ['pearson_coeff','p-value'], columns = ['result'])

result
pearson_coeff  0.080954
p-value        0.000000

_Avec une p-value de 0 et un coefficient > 5%, on observe une forte corrélation entre les variables "events_transaction" et "User frequency", ce qui semble logique compte tenu des varaibles considérées._

* Si on réalise la même étude variable cible (events_transaction) et "Product frequency", on observe une corrélation beaucoup moins marquée car le coefficient est supérieur à 1 et la p-value s'approche de 5%. On peut tout de même considérer qu'il y a corrélation :

In [44]:
pd.DataFrame(pearsonr(df_events['events_transaction'],df_events['Product frequency']), index = ['pearson_coeff','p-value'], columns = ['result'])

result
pearson_coeff   1.623410e-02
p-value        5.617452e-160

Si l'on cherche à croiser la variable cible (events_transaction) et "events_view", on observe avec surprise un coefficient négatif. Cela nous indique que les variables, bien que corrélées par p-value évoluent en sens opposé :

In [45]:
pd.DataFrame(pearsonr(df_events['events_transaction'],df_events['events_view']), index = ['pearson_coeff','p-value'], columns = ['result'])

result
pearson_coeff -0.489326
p-value        0.000000

En croisant la variable cible (events_transaction) et "events_addtocart", on observe que les variables sont peu corrélées et évoluent en sens opposé, indiquant un taux d'abandon de panier très élevé et une rupture très marquée dans le parcours d'achat :

In [46]:
pd.DataFrame(pearsonr(df_events['events_transaction'],df_events['events_addtocart']), index = ['pearson_coeff','p-value'], columns = ['result'])

result
pearson_coeff  -1.452283e-02
p-value        1.994777e-128

* L'utilisation de la fonction crosstab est utile pour identifier une corrélation entre deux variables quantitatives. Elle permet de croiser les différentes catégories de deux variables. On créer ici la table de contingence entre la variable cible events_transaction et User category :

In [47]:
table = pd.crosstab(df_events['events_transaction'],df_events['User category'])
table

User category       Infrequent user  Recurrent user  Big user   Bots
events_transaction                                                  
0                           2200007          326965    192297  13915
1                              9268            5740      6255   1194

* Puis on réalise un test de proportions entre l'effectif de la cellule et l'effectif total de la colonne, ici le χ2 sur la table de contingence déterminée précédemment (variable cible events_transaction et day of week)

In [48]:
from scipy.stats import chi2_contingency

resultats_test = chi2_contingency(table)
statistique = resultats_test[0]
p_valeur = resultats_test[1]
degre_liberte = resultats_test[2]

print(statistique,p_valeur,degre_liberte)

30471.797136536483 0.0 3


* Un dernier test V Cramer nous permet de voir la corrélation entre les deux variables :

In [49]:
def V_Cramer (table,N):
    stat_chi2 = chi2_contingency(table)[0]
    k = table.shape[0]
    r = table.shape[1]
    phi = max(0,(stat_chi2/N)-((k-1)*(r-1)/(N-1)))
    k_corr = k - (np.square(k-1)/(N-1))
    r_corr = r - (np.square(r-1)/(N-1))
    return np.sqrt(phi/min(k_corr - 1,r_corr - 1))

V_Cramer(table,df_events.shape[0])

0.10515172812105962

* On observe que le V de Cramer donne un résultat relativement proche de 0, donc il existe une corrélation entre les variables events_transaction et day_of_week. Si on regarde la table de contingence on remarque que les jours de semaine il y a plus de ventes que le Week-end. Le jour avec le nombre de transactions le plus élevé est le mercredi. 

* On réalise les mêmes manipulations statistiques avec d'autres varaibles et la variable cible. Table de contingence entre les variable events_view et Product category :

In [50]:
table = pd.crosstab(df_events['events_transaction'],df_events['Product category'])
table

Product category    Unpopular products  Popular products  \
events_transaction                                         
0                              1894312            788112   
1                                13307              8528   

Product category    Very popular products  Most popular products  
events_transaction                                                
0                                   28955                  13217  
1                                     387                    102

* Puis on teste le χ2 sur la table de contingence déterminée précédemment (variables events_transaction et Product category)



In [51]:
resultats_test = chi2_contingency(table)
statistique = resultats_test[0]
p_valeur = resultats_test[1]
degre_liberte = resultats_test[2]

print(statistique,p_valeur,degre_liberte)

1064.1184072898184 2.215472700205196e-230 3


* Enfin on réalise un test de V_Cramer par rapport au tableau de contingence (variables events_transaction et Product category) :

In [52]:
def V_Cramer (table,N):
    stat_chi2 = chi2_contingency(table)[0]
    k = table.shape[0]
    r = table.shape[1]
    phi = max(0,(stat_chi2/N)-((k-1)*(r-1)/(N-1)))
    k_corr = k - (np.square(k-1)/(N-1))
    r_corr = r - (np.square(r-1)/(N-1))
    return np.sqrt(phi/min(k_corr - 1,r_corr - 1))

V_Cramer(table,df_events.shape[0])

0.019623238570507805

* On observe que le V de Cramer donne également un résultat relativement proche de 0, donc il existe une corrélation entre les variables events_transaction et Product category. En observant la table de contingence on remarque que les produits qui gènèrent le plus d'événements ne sont pas ceux qui se vendent le plus. Ce sont les produits de la catégorie 'Very popular products' qui ont le pourcentage d'événements de vente le plus important.

* On créer un nouveau dataframe qui va être utilisé pour le clustering et qui marque pour chaque visitorid le nombre de fois où il a un événement de type nombre de vies, nombre d'ajouts au panier et de transactions :